[GH-1754] Feature: card badges

kamre opened the pull request:
#### Summary
Badges with information about description/comments/checkboxes for cards on kanban and gallery views.

Looks like this:
![image](https://user-images.githubusercontent.com/5587620/147684547-701ed044-112d-4ab0-91ac-41a4f4557e97.png)

Not visible by default because it takes additional space on the card. Can be shown/hidden via the view menu:
![image](https://user-images.githubusercontent.com/5587620/147684733-2b01ad2c-a736-461c-84d6-f423b62fbb81.png)

Corresponding code changes:
* `CardBadges` component added and used in `GalleryCard` and `KanbanCard`
* optional field `cardBadgesVisible` added to `BoardViewFields`
* action for show/hide badges added to `ViewHeaderActionsMenu` for kanban and gallery views
* `MessageIcon` component for corresponding Compass icon added
* unit and Cypress tests added
* minor improvements for `MenuWrapper` and `TestBlockFactory`


#### Ticket Link
Fixes #1754 

@asaadmahmood please review from UI/UX side

mattermod commented:
Hello @kamre,

Thanks for your pull request! A Core Committer will review your pull request soon. For code contributions, you can learn more about the review process [here](https://developers.mattermost.com/contribute/getting-started/code-review/).

jespino commented on webapp/src/components/cardBadges.tsx:
```diff
@@ -0,0 +1,107 @@
+    const intl = useIntl()
+    const {checkboxes} = badges
+    return (
+        <div className={`cardBadges ${className || ''}`}>
```
Tiny detail, by convention we use the component name for the root component className inside the component. In this case should be `CardBadges` with uppercase at the beginning.

kamre replied:
Good to know. Fixed.

jespino reviewed (changes requested):
I added a comment about the name of the class name, other than that looks good to me. Great work @kamre! :tada:

kamre commented on webapp/src/components/properties/lastModifiedAt/__snapshots__/lastModifiedAt.test.tsx.snap:
```diff
@@ -5,7 +5,7 @@ exports[`componnets/properties/lastModifiedAt should match snapshot 1`] = `
     class="LastModifiedAt octo-propertyvalue readonly"
   >
-    June 15, 4:22 PM
+    June 15, 2021, 4:22 PM
```
Rendering of the predefined date in the test has changed with the arrival of the new year.

jespino replied:
Interesting.

jespino reviewed (approved):
Now looks good to me. Thanks @kamre! :confetti_ball:

sbishel reviewed (approved):
Code looks good. I just wanted to make sure we have UX (@asaadmahmood ) and PM 
 (@wuwinson) review placement of the 'Show/Hide card badges' MenuItem. (Properties vs Ellipsis Menu).

I have this PR installed at 
https://sbishel.test.mattermost.cloud/
If you want to play with it.

asaadmahmood reviewed (commented):
@kamre I think this should actually exist under the Properties as a toggle.

kamre commented:
@asaadmahmood I agree that from the point of view on toggling visible elements on the cards it is related to `Properties` menu, but it is not related to card properties themselves currently because only card content blocks are used.

Do you think that adding `Card badges` or even just `Badges` to `Properties` menu at the bottom will be a better solution?

asaadmahmood commented:
@kamre Yup, those may not strictly be card properties, but we can consider them properties in a general sense.
With regards to the term Badges, or Card Badges, not sure.
I think we should just say "Comments" and "Description"

If we're doing something for the description, I honestly am not sure about the checkboxes thing, as that's only part of the description, and I'm not sure that would be isolated as a different item for just the checkboxes.

CC @wuwinson @chenilim

wuwinson commented:
@kamre @asaadmahmood Agreed that we should move the toggles to the properties section for now. Would keep it simple with a single toggle for all badges. Current location is not very discoverable and don't think this feature will get much utilization if it stays there.

Not sure how helpful the "Description" badge is, but think we should at least keep the "Comments" and "Checkbox" badges since it provides a bit more helpful info without requiring the user to click into the card. Okay with keeping all 3, but would cut the "Description" one if anything. Thoughts? cc @chenilim

chenilim commented:
Thanks @kamre! This is a great feature! I agree with moving the toggle to under the Properties menu (conceptually, it's what to display on the card preview) - I suppose this could apply to Boards, Calendar, and Gallery. For tables, it could show up as another column, but that might be a future enhancement.

One minor thing is the "Text" icon appears larger than the others (comments and checkbox). It also looks similar to a grab-handle.

<img width="277" alt="Card badges" src="https://user-images.githubusercontent.com/46905241/148428443-12a0193c-3dea-4d5e-9e9b-1facde860b83.png">

This is something we can iterate on, but perhaps there's a better icon to represent the description? cc @asaadmahmood.

In any case, I'm in favor of checking this in and iterating - now that v0.12 has branched, we have time to make updates.

wuwinson reviewed (approved):
Sounds good @chenilim!

kamre commented:
@chenilim @asaadmahmood I have moved show/hide for card badges to `Properties` menu:
![image](https://user-images.githubusercontent.com/5587620/148602902-ac3160db-981d-4155-9922-c97c43a78420.png)
Something similar was already done for the card title in the gallery view when non-property is added to this menu.

Also I have added card badges support to calendar view. Nothing for table view yet, this can be done separately.

Regarding the icons: I have just reused ones from the context menu for `Add content` in the card dialog:
![image](https://user-images.githubusercontent.com/5587620/148603431-6166f60b-6732-4c11-a318-03792cbc042a.png)
They are currently just inline svg and not Compass icons. But for comments I have used the icon from Compass set.

So if there are better icons in the Compass set for text and checkbox we can use them.

sbishel commented:
/update-branch

kamre commented:
@sbishel I have noticed that Cypress test is failing after the merge. Something is wrong with adding comments, will look into this tomorrow.

sbishel commented:
Thanks @kamre, just ping me and I'll get it merged.

kamre commented:
@sbishel fixed Cypress test: used explicit `blur` event after typing comment text before clicking on `Send` button. Manually this issue can't be reproduced so probably it is Cypress issue with changing focus to the button and not triggering blur event.

sbishel commented:
Awesome...thanks @kamre
